File: golang/dance.go
package main

import (
	"math"
	"github.com/fogleman/gg"
)

type Planet int

const (
	Mercury Planet = iota
	Venus
	Earth
	Mars
	Jupiter
	Saturn
	Uranus
	Neptune
	Pluto
)

func nameOfPlanet(planet Planet) string {
	switch planet {
	case Mercury:
		return "Mercury"
	case Venus:
		return "Venus"
	case Earth:
		return "Earth"
	case Mars:
		return "Mars"
	case Jupiter:
		return "Jupiter"
	case Saturn:
		return "Saturn"
	case Uranus:
		return "Uranus"
	case Neptune:
		return "Neptune"
	case Pluto:
		return "Pluto"
	default:
		//TODO: return error
		return ""
	}
}

func orbitOfPlanet(planet Planet) float64 {
	switch planet {
	case Mercury:
		return 57.91
	case Venus:
		return 108.21
	case Earth:
		return 149.60
	case Mars:
		return 227.92
	case Jupiter:
		return 778.57
	case Saturn:
		return 1433.5
	case Uranus:
		return 2872.46
	case Neptune:
		return 4495.1
	case Pluto:
		return 5869.7
	default:
		//TODO: return error
		return 1.0
	}
}

func yearOfPlanet(planet Planet) float64 {
	switch planet {
	case Mercury:
		return 87.969
	case Venus:
		return 224.701
	case Earth:
		return 365.256
	case Mars:
		return 686.980
	case Jupiter:
		return 4332.6
	case Saturn:
		return 10759.2
	case Uranus:
		return 30685.0
	case Neptune:
		return 60190.0
	case Pluto:
		return 60190.0
	default:
		//TODO: return error
		return 1.0
	}
}

func colorOfI(i int) (float64, float64, float64, float64) {
	alpha := 1.0
	switch i {
	case 0:
		return 0.0, 0.0, 0.0, alpha
	case 1:
		return 0.0, 0.0, 1.0, alpha
	case 2:
		return 1.0, 0.0, 0.0, alpha
	case 3:
		return 0.0, float64(0xFF) / 255.0, 0.0, alpha
	case 4:
		return float64(0x80) / 255.0, 0.0, float64(0x80) / 255.0, alpha
	case 5:
		return float64(0x80) / 255.0, 0.0, 0.0, alpha
	case 6:
		return float64(0x8B) / 255.0, 0.0, 0.0, alpha
	case 7:
		return float64(0x20) / 255.0, float64(0x2F) / 255.0, float64(0x55) / 255.0, alpha
	default:
		return 1.0, float64(0xA5) / 255.0, 0.0, alpha
	}
}

func dance(dc *gg.Context, canvasLen int, innerPlanet Planet, outerPlanet Planet, orbits float64) {
	innerPlanetYear := yearOfPlanet(innerPlanet)
	outerPlanetYear := yearOfPlanet(outerPlanet)
	innerPlanetRadius := orbitOfPlanet(innerPlanet)
	outerPlanetRadius := orbitOfPlanet(outerPlanet)
	intervalDays := outerPlanetYear / 75
	xcenter := float64(canvasLen / 2)
	ycenter := float64(canvasLen / 2)
	r1 := float64(ycenter)
	r2 := float64(r1) * innerPlanetRadius / outerPlanetRadius
	r := 0.0
	a1 := 0.0
	a2 := 0.0
	rstop := outerPlanetYear * orbits
	a1Interval := 2 * math.Pi * intervalDays / outerPlanetYear
	a2Interval := 2 * math.Pi * intervalDays / innerPlanetYear
	dc.SetRGB(1.0, 1.0, 1.0)
	dc.DrawRectangle(0.0, 0.0, float64(canvasLen), float64(canvasLen))
	dc.Fill()
	for r < rstop {
		i := int(r / intervalDays / 75.0)
		//		fmt.Println("i ", i)
		a1 = a1 - a1Interval
		a2 = a2 - a2Interval
		x1 := r1*math.Cos(a1) + xcenter
		y1 := r1*math.Sin(a1) + ycenter
		x2 := r2*math.Cos(a2) + xcenter
		y2 := r2*math.Sin(a2) + ycenter
		dc.SetRGBA(colorOfI(i))
		dc.DrawLine(x1, y1, x2, y2)
		dc.Stroke()
		r = r + intervalDays
	}
}

func main() {
	const canvasLen = 400
	dc := gg.NewContext(canvasLen, canvasLen)
	dance(dc, canvasLen, Venus, Earth, 8.0)
	dc.SavePNG("result.png")
}

File: rust/src/dance.rs
extern crate gdk;
extern crate gio;
extern crate gtk;
extern crate cairo;

use gio::prelude::*;
use gtk::prelude::*;

use std::env::args;

// gtk doc
// http://gtk-rs.org/docs/gtk/

fn color_of_i(i: i32) -> (f64, f64, f64) {
    match i {
        0 => { return (0.0, 0.0, 0.0); }
        1 => { return (0.0, 0.0, 1.0); }
        2 => { return (1.0, 0.0, 0.0); }
        3 => { return (0.0, 1.0, 0.0); }
        4 => { return (0x80 as f64/0xFF as f64, 0.0, 0x80 as f64/0xFF as f64); }
        5 => { return (0x80 as f64/0xFF as f64, 0.0, 0.0); }
        6 => { return (0.0, 0.0, 0x8B as f64/0xFF as f64); }
        7 => { return (0x8B as f64/0xFF as f64, 0.0, 0.0); }
        _ => { return (1.0, 0xA5 as f64/0xFF as f64, 0.0); }
    }
}

fn dance(cairo: &cairo::Context, orbits: f64){
    //TODO: get as argument
    let pi =  std::f64::consts::PI;
    let canvas_len = 400;
    //earth, venus
    let outer_planet_year = 365.256;
    let inner_planet_year = 224.701;
    let outer_planet_radius = 149.60;
    let inner_planet_radius = 108.21;
    let interval_days = outer_planet_year / 75.0;
    let ycenter = (canvas_len / 2) as f64;
    let xcenter = ycenter;
    let r1 = ycenter;
    let r2 = r1 * inner_planet_radius / outer_planet_radius;
    let mut r = 0.0;
    let rstop = outer_planet_year * orbits;
    let mut a1 = 0.0;
    let a1_interval = 2. * pi * interval_days / outer_planet_year;
    let mut a2 = 0.0;
    let a2_interval = 2. * pi * interval_days / inner_planet_year;

    //fill white background
    cairo.set_source_rgb(1.0, 1.0, 1.0);
    cairo.rectangle(0.0, 0.0, canvas_len as f64, canvas_len as f64);
    cairo.fill();

    while r < rstop {
        let i = (r / interval_days / 75.).floor() as i32;
        let color = color_of_i(i);
        //TODO: ???
        cairo.set_source_rgb(color.0, color.1, color.2);
        a1 -= a1_interval;
        a2 -= a2_interval;
        let x1 = r1 * a1.cos() + xcenter;
        let y1 = r1 * a1.sin() + ycenter;
        let x2 = r2 * a2.cos() + xcenter;
        let y2 = r2 * a2.sin() + ycenter;
        //draw line
        cairo.move_to(x1, y1);
        cairo.line_to(x2, y2);
        cairo.stroke();
        r += interval_days
    }
}

fn build_ui(application: &gtk::Application) {
    let canvas_len = 400;
    let window = gtk::ApplicationWindow::new(application);

    window.set_title("First GTK+ Program");
    window.set_border_width(10);
    window.set_position(gtk::WindowPosition::Center);
    window.set_default_size(canvas_len, canvas_len);
    let area = gtk::DrawingArea::new();
    area.set_size_request(400, 400);

    //    let button = gtk::Canvas
    window.add(&area);
    area.connect_draw(|_, cairo| {
        let orbits = 8.0;
        dance(cairo, orbits);
        Inhibit(false)
    });
    window.show_all();
}

fn main() {
    let application =
        gtk::Application::new(Some("com.github.mamewo.dance_of_the_plantes"), Default::default())
            .expect("Initialization failed...");

    application.connect_activate(|app| {
        build_ui(app);
    });

    application.run(&args().collect::<Vec<_>>());
}
